packages/compass-schema-validation/src/modules/validation.ts

```diff
@@ -385,40 +446,55 @@ export const syntaxErrorOccurred = (
 export const fetchValidation = (namespace: {
   database: string;
   collection: string;
-}): SchemaValidationThunkAction<void> => {
-  return (dispatch, _getState, { dataService }) => {
-    dataService.collectionInfo(namespace.database, namespace.collection).then(
-      (collInfo) => {
-        const validation = validationFromCollection(null, collInfo ?? {});
-
-        if (!validation.validator) {
-          const newValidation = { ...validation, validator: '{}' };
-          dispatch(validationFetched(newValidation));
-          dispatch(isLoadedChanged(true));
-          return;
-        }
-
-        // TODO(COMPASS-4989): EJSON??
-        const newValidation = {
-          ...validation,
-          validator: EJSON.stringify(validation.validator, undefined, 2),
-        };
-
-        dispatch(validationFetched(newValidation));
-        dispatch(zeroStateChanged(false));
-        dispatch(isLoadedChanged(true));
-      },
-      (err: Error) => {
+}): SchemaValidationThunkAction<Promise<void>> => {
+  return async (dispatch, _getState, { dataService, preferences }) => {
+    try {
+      const collInfo = await dataService.collectionInfo(
+        namespace.database,
+        namespace.collection
+      );
+      if (!collInfo?.validation?.validator) {
         dispatch(
-          validationFetched({
-            ...validationFromCollection(err),
-            validator: undefined,
-          })
+          setValidationToDefault(
```

Comment:
This is still validation fetched in terms of what action actually happened in the app. It's strongily recommended for reducer to handle the logic of [how to map this to the state](https://redux.js.org/style-guide#reducers-should-own-the-state-shape)

I'd also maybe suggest to get rid of bulk dispatch here while you're at it as this is an easy clean-up that you can do drive-by, but I'll leave it to you

Reply:
Removed the bulk dispatches. Updated the reducer here and some of the actions to be have a bit less indirection. Half of the logic was happening here and then more was happening in the reducer, which was used by two separate actions that weren't really related. Let me know if the changes make sense. I was trying to avoid making too large of changes here, but I think it might help in the long run.